src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+package ch.cern.eam.wshub.core.services.administration.impl;
+
+import ch.cern.eam.wshub.core.client.InforContext;
+import ch.cern.eam.wshub.core.services.administration.UserGroupMenuService;
+import ch.cern.eam.wshub.core.tools.ApplicationData;
+import ch.cern.eam.wshub.core.tools.InforException;
+import ch.cern.eam.wshub.core.tools.Tools;
+import net.datastream.schemas.mp_entities.extmenus_001.ExtMenus;
+import net.datastream.schemas.mp_entities.extmenushierarchy_001.ExtMenusHierarchy;
+import net.datastream.schemas.mp_fields.*;
+import net.datastream.schemas.mp_functions.mp6005_001.MP6005_GetExtMenusHierarchy_001;
+import net.datastream.schemas.mp_functions.mp6043_001.MP6043_AddExtMenus_001;
+import net.datastream.schemas.mp_functions.mp6045_001.MP6045_DeleteExtMenus_001;
+import net.datastream.wsdls.inforws.InforWebServicesPT;
+
+import java.util.*;
+
+public class UserGroupMenuServiceImpl implements UserGroupMenuService {
+    private Tools tools;
+    private InforWebServicesPT inforws;
+    private ApplicationData applicationData;
+
+    public UserGroupMenuServiceImpl(ApplicationData applicationData, Tools tools, InforWebServicesPT inforWebServicesToolkitClient) {
+        this.applicationData = applicationData;
+        this.tools = tools;
+        this.inforws = inforWebServicesToolkitClient;
+    }
+
+    private void addFolderToEntries(List<GenericMenuEntry> entries, FOLDER_Type parent) {
+        entries.add(new GenericMenuEntry(parent));
+        for(FOLDER_Type childFolder : parent.getFOLDER()) {
+            addFolderToEntries(entries, childFolder);
+        }
+        for (FUNCTION_Type childFunction : parent.getFUNCTION()) {
+            addFunctionToEntries(entries, childFunction);
+        }
+    }
+
+    private void addFunctionToEntries(List<GenericMenuEntry> entries, FUNCTION_Type parent) {
+        entries.add(new GenericMenuEntry(parent));
+    }
+
+    private class GenericMenuEntry {
+        private String id;
+        private String description;
+        private String parent;
+        private String functionId;
+
+        public GenericMenuEntry(FOLDER_Type folder) {
+            this.id = folder.getEXTMENUCODE();
+            this.description = folder.getFOLDERID().getFOLDERDESCRIPTION();
+            this.parent = folder.getEXTMENUPARENT();
+            this.functionId = folder.getFOLDERID().getFOLDERCODE();
+        }
+
+        public GenericMenuEntry(MENU_Type menu) {
+            this.id = menu.getEXTMENUCODE();
+            this.description = menu.getMENUID().getMENUDESCRIPTION();
```

Comment:
Potential null pointer on a constructor. Validation should be on this side as opposed to each time constructor is invoked

Reply:
Solved.